### ISSUE TITLE
`react-native autolink-windows` has regressed in speed since 0.65

### ISSUE THREAD

**Issue by tido64** [microsoft/react-native-windows#9518]
### Problem Description

`react-native autolink-windows` has regressed in speed since 0.65

| Version |    Time |
| :------ | ------: |
| 0.64    |  3639ms |
| 0.65    | 17793ms |
| 0.66    | 22134ms |
| 0.68    | 22488ms |

### Steps To Reproduce

n/a

### Expected Results

`react-native autolink-windows` should take as long as 0.64 did.

### CLI version

5.0.1

### Environment

```markdown
System:
    OS: Windows 10 10.0.22000
    CPU: (32) x64 AMD Ryzen Threadripper PRO 3955WX 16-Cores
    Memory: 117.20 GB / 127.86 GB
  Binaries:
    Node: 14.17.5 - C:\Program Files\nodejs\node.EXE
    Yarn: 3.1.1 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.4.1 - D:\react-native-test-app\node_modules\.bin\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.11.32126.315 (Visual Studio Enterprise 2019)
  Languages:
    Java: 11.0.14 - /c/Program Files/Eclipse Adoptium/jdk-11.0.14.9-hotspot/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.1 => 17.0.1
    react-native: ^0.64.3 => 0.64.3
    react-native-windows: ^0.64.25 => 0.64.27
  npmGlobalPackages:
    *react-native*: Not Found
```

### Target Device(s)

Desktop

### Visual Studio Version

Visual Studio 2019

### Build Configuration

Debug

**Comment by tido64** [comment]
cc @NickGerleman

**Comment by NickGerleman** [comment]
@chrisglein this seems like a pretty clear regression to build performance. Could you help me understand why you considered this item an enhancement?

**Comment by NickGerleman** [comment]
cc @Khalef1

@tido64 created this item as a bug, and it is pretty clearly one as well in my eyes. I would appreciate it if we didn't change the labels our engineers use when they file these. Especially given it feeds into tooling used to understand quality.

**Comment by jonthysell** [comment]
Off the top of my head I can't think of any changes in 0.65 that would account for this. I'll look into it.

**Comment by jonthysell** [comment]
Ok, so you have 0 dependencies? Can you post the result of `react-native config`, and about how long that takes to run?

Also, what does it look like when you run `run-windows`? Are you still getting those large run-times?

**Comment by tido64** [comment]
Yup, 0 dependencies. For `run-windows`, if the app is already built, auto-linking dominates the time spent.

Here's the output for `run-windows`:

```
✔ Restoring NuGet packages
 ✔ Auto-linking...
Success: No auto-linking changes necessary. (19824ms)
 ✔ Found Solution: D:\react-native-test-app\example\windows\Example.sln
 ℹ Build configuration: Debug
 ℹ Build platform: x64
 ✔ Building Solution
 ✔ Starting the React-Native Server
 ✔ Enabling Developer Mode
 ✔ Deploying D:\react-native-test-app\example\windows\x64\Debug\ReactTestApp\ReactTestApp.build.appxrecipe
 ✔ Verifying loopbackExempt
 ✔ Starting the app
```

And for `react-native config`:

```
$ time yarn react-native config
{
  "root": "D:\\react-native-test-app\\example",
  "reactNativePath": "D:\\react-native-test-app\\example\\node_modules\\react-native",
  "dependencies": {
    "react-native-test-app": {
      "root": "D:\\react-native-test-app",
      "name": "react-native-test-app",
      "platforms": {
        "ios": {
          "sourceDir": "D:\\react-native-test-app\\example\\ios",
          "folder": "D:\\react-native-test-app",
          "pbxprojPath": "D:\\react-native-test-app\\ios\\ReactTestApp.xcodeproj\\project.pbxproj",
          "podfile": "D:\\react-native-test-app\\example\\ios\\Podfile",
          "podspecPath": "D:\\react-native-test-app\\ReactTestApp-DevSupport.podspec",
          "projectPath": "D:\\react-native-test-app\\ios\\ReactTestApp.xcodeproj",
          "projectName": "ReactTestApp.xcodeproj",
          "libraryFolder": "Libraries",
          "sharedLibraries": [],
          "plist": [],
          "scriptPhases": [],
          "configurations": []
        },
        "android": null,
        "windows": null
      },
      "assets": [],
      "hooks": {},
      "params": []
    }
  },
  "commands": [
    {
      "name": "log-ios",
      "description": "starts iOS device syslog tail"
    },
    {
      "name": "run-ios",
      "description": "builds your app and starts it on iOS simulator",
      "examples": [
        {
          "desc": "Run on a different simulator, e.g. iPhone SE (2nd generation)",
          "cmd": "react-native run-ios --simulator \"iPhone SE (2nd generation)\""
        },
        {
          "desc": "Pass a non-standard location of iOS directory",
          "cmd": "react-native run-ios --project-path \"./app/ios\""
        },
        {
          "desc": "Run on a connected device, e.g. Max's iPhone",
          "cmd": "react-native run-ios --device \"Max's iPhone\""
        },
        {
          "desc": "Run on the AppleTV simulator",
          "cmd": "react-native run-ios --simulator \"Apple TV\"  --scheme \"helloworld-tvOS\""
        }
      ],
      "options": [
        {
          "name": "--simulator <string>",
          "description": "Explicitly set simulator to use. Optionally include iOS version between parenthesis at the end to match an exact version: \"iPhone 6 (10.0)\"",
          "default": "iPhone 13"
        },
        {
          "name": "--configuration <string>",
          "description": "Explicitly set the scheme configuration to use",
          "default": "Debug"
        },
        {
          "name": "--scheme <string>",
          "description": "Explicitly set Xcode scheme to use"
        },
        {
          "name": "--project-path <string>",
          "description": "Path relative to project root where the Xcode project (.xcodeproj) lives.",
          "default": "ios"
        },
        {
          "name": "--device [string]",
          "description": "Explicitly set device to use by name.  The value is not required if you have a single device connected."
        },
        {
          "name": "--udid <string>",
          "description": "Explicitly set device to use by udid"
        },
        {
          "name": "--no-packager",
          "description": "Do not launch packager while building"
        },
        {
          "name": "--verbose",
          "description": "Do not use xcbeautify or xcpretty even if installed"
        },
        {
          "name": "--port <number>",
          "default": 8081
        },
        {
          "name": "--terminal <string>",
          "description": "Launches the Metro Bundler in a new window using the specified terminal path."
        }
      ]
    },
    {
      "name": "log-android",
      "description": "starts logkitty"
    },
    {
      "name": "run-android",
      "description": "builds your app and starts it on a connected Android emulator or device",
      "options": [
        {
          "name": "--root <string>",
          "description": "[DEPRECATED - root is discovered automatically] Override the root directory for the android build (which contains the android directory)",
          "default": ""
        },
        {
          "name": "--variant <string>",
          "description": "Specify your app's build variant",
          "default": "debug"
        },
        {
          "name": "--appFolder <string>",
          "description": "[DEPRECATED – use \"project.android.appName\" in react-native.config.js] Specify a different application folder name for the android source. If not, we assume is \"app\""
        },
        {
          "name": "--appId <string>",
          "description": "Specify an applicationId to launch after build. If not specified, `package` from AndroidManifest.xml will be used.",
          "default": ""
        },
        {
          "name": "--appIdSuffix <string>",
          "description": "Specify an applicationIdSuffix to launch after build.",
          "default": ""
        },
        {
          "name": "--main-activity <string>",
          "description": "Name of the activity to start",
          "default": "MainActivity"
        },
        {
          "name": "--deviceId <string>",
          "description": "builds your app and starts it on a specific device/simulator with the given device id (listed by running \"adb devices\" on the command line)."
        },
        {
          "name": "--no-packager",
          "description": "Do not launch packager while building"
        },
        {
          "name": "--port <number>",
          "default": 8081
        },
        {
          "name": "--terminal <string>",
          "description": "Launches the Metro Bundler in a new window using the specified terminal path.",
          "default": "xterm-256color"
        },
        {
          "name": "--tasks <list>",
          "description": "Run custom Gradle tasks. By default it's \"installDebug\""
        },
        {
          "name": "--no-jetifier",
          "description": "Do not run \"jetifier\" – the AndroidX transition tool. By default it runs before Gradle to ease working with libraries that don't support AndroidX yet. See more at: https://www.npmjs.com/package/jetifier."
        },
        {
          "name": "--active-arch-only",
          "description": "Build native libraries only for the current device architecture for debug builds.",
          "default": false
        }
      ]
    },
    {
      "name": "init-test-app",
      "description": "Initializes a new test app project",
      "options": [
        {
          "name": "--destination [string]",
          "description": "Path to the directory where the test app should be created",
          "default": "D:\\react-native-test-app\\example\\test-app"
        },
        {
          "name": "--name [string]",
          "description": "Display name of the test app",
          "default": "example-test-app"
        },
        {
          "name": "--platform [string]",
          "description": "Specific platform to support: all, android, ios, macos, windows",
          "default": "all"
        }
      ]
    },
    {
      "name": "autolink-windows",
      "description": "performs autolinking",
      "options": [
        {
          "name": "--logging",
          "description": "Verbose output logging"
        },
        {
          "name": "--check",
          "description": "Only check whether any autolinked files need to change"
        },
        {
          "name": "--sln [string]",
          "description": "Override the app solution file determined by 'react-native config', e.g. windows\\myApp.sln"
        },
        {
          "name": "--proj [string]",
          "description": "Override the app project file determined by 'react-native config', e.g. windows\\myApp\\myApp.vcxproj"
        },
        {
          "name": "--no-telemetry",
          "description": "Disables sending telemetry that allows analysis of usage and failures of the react-native-windows CLI"
        }
      ]
    },
    {
      "name": "run-windows",
      "description": "builds your app and starts it on a connected Windows desktop, emulator or device",
      "options": [
        {
          "name": "--release",
          "description": "Specifies a release build"
        },
        {
          "name": "--root [string]",
          "description": "Override the root directory for the windows build which contains the windows folder."
        },
        {
          "name": "--arch [string]",
          "description": "The build architecture (ARM64, x86, x64)",
          "default": "x64"
        },
        {
          "name": "--singleproc",
          "description": "Disables multi-proc build"
        },
        {
          "name": "--emulator",
          "description": "Deploys the app to an emulator"
        },
        {
          "name": "--device",
          "description": "Deploys the app to a connected device"
        },
        {
          "name": "--target [string]",
          "description": "Deploys the app to the specified GUID for a device."
        },
        {
          "name": "--remote-debugging",
          "description": "Deploys the app in remote debugging mode."
        },
        {
          "name": "--logging",
          "description": "Enables logging"
        },
        {
          "name": "--no-packager",
          "description": "Do not launch packager while building"
        },
        {
          "name": "--bundle",
          "description": "Enable Bundle configuration and it would be ReleaseBundle/DebugBundle other than Release/Debug"
        },
        {
          "name": "--no-launch",
          "description": "Do not launch the app after deployment"
        },
        {
          "name": "--no-autolink",
          "description": "Do not run autolinking"
        },
        {
          "name": "--no-build",
          "description": "Do not build the solution"
        },
        {
          "name": "--no-deploy",
          "description": "Do not deploy the app"
        },
        {
          "name": "--deploy-from-layout",
          "description": "Force deploy from layout"
        },
        {
          "name": "--sln [string]",
          "description": "Override the app solution file determined by 'react-native config', e.g. windows\\myApp.sln"
        },
        {
          "name": "--proj [string]",
          "description": "Override the app project file determined by 'react-native config', e.g. windows\\myApp\\myApp.vcxproj"
        },
        {
          "name": "--msbuildprops [string]",
          "description": "Comma separated props to pass to msbuild, eg: prop1=value1,prop2=value2"
        },
        {
          "name": "--buildLogDirectory [string]",
          "description": "Optional directory where msbuild log files should be stored"
        },
        {
          "name": "--info",
          "description": "Dump environment information"
        },
        {
          "name": "--direct-debugging [number]",
          "description": "Enable direct debugging on specified port"
        },
        {
          "name": "--no-telemetry",
          "description": "Disables sending telemetry that allows analysis of usage and failures of the react-native-windows CLI"
        }
      ]
    }
  ],
  "assets": [],
  "healthChecks": [
    {
      "label": "Windows",
      "healthchecks": [
        {
          "label": "Free space on D: > 15 GB"
        },
        {
          "label": "Installed memory >= 16 GB"
        },
        {
          "label": "Windows version > 10.0.16299.0"
        },
        {
          "label": "Developer mode is on"
        },
        {
          "label": "Long path support is enabled"
        },
        {
          "label": "Choco"
        },
        {
          "label": "git"
        },
        {
          "label": "Compilers, build tools, SDKs and Visual Studio"
        },
        {
          "label": "NodeJS LTS"
        },
        {
          "label": "Chrome"
        },
        {
          "label": "Yarn"
        },
        {
          "label": ".net core 3.1"
        }
      ]
    }
  ],
  "platforms": {
    "ios": {},
    "android": {},
    "windows": {
      "npmPackageName": "react-native-windows"
    }
  },
  "project": {
    "ios": {
      "sourceDir": "D:\\react-native-test-app\\example\\ios",
      "folder": "D:\\react-native-test-app\\example",
      "pbxprojPath": "D:\\react-native-test-app\\example\\node_modules\\.generated\\ios\\ReactTestApp.xcodeproj\\project.pbxproj",
      "podfile": "D:\\react-native-test-app\\example\\ios\\Podfile",
      "podspecPath": "D:\\react-native-test-app\\example\\Example-Tests.podspec",
      "projectPath": "D:\\react-native-test-app\\example\\node_modules\\.generated\\ios\\ReactTestApp.xcodeproj",
      "projectName": "ReactTestApp.xcodeproj",
      "libraryFolder": "Libraries",
      "sharedLibraries": [],
      "plist": [],
      "scriptPhases": []
    },
    "android": {
      "sourceDir": "D:\\react-native-test-app\\example\\android",
      "isFlat": false,
      "folder": "D:\\react-native-test-app\\example",
      "stringsPath": "D:\\react-native-test-app\\example\\android\\src\\main\\res\\values\\strings.xml",
      "manifestPath": "D:\\react-native-test-app\\android\\app\\src\\main\\AndroidManifest.xml",
      "buildGradlePath": "D:\\react-native-test-app\\example\\android\\build.gradle",
      "settingsGradlePath": "D:\\react-native-test-app\\example\\android\\settings.gradle",
      "assetsPath": "D:\\react-native-test-app\\example\\android\\src\\main\\assets",
      "mainFilePath": "D:\\react-native-test-app\\example\\android\\src\\main\\java\\com\\microsoft\\reacttestapp\\MainApplication.java",
      "packageName": "com.microsoft.reacttestapp",
      "packageFolder": "com\\microsoft\\reacttestapp",
      "appName": ""
    },
    "windows": {
      "folder": "D:\\react-native-test-app\\example",
      "sourceDir": "windows",
      "solutionFile": "Example.sln",
      "project": {
        "projectFile": "..\\node_modules\\.generated\\windows\\ReactTestApp\\ReactTestApp.vcxproj",
        "projectName": "ReactTestApp",
        "projectLang": "cpp",
        "projectGuid": "{b44cead7-fbff-4a17-95ea-ff5434bbd79d}"
      },
      "experimentalFeatures": {
        "UseHermes": "false",
        "UseWinUI3": "false",
        "UseExperimentalNuget": "false"
      }
    }
  }
}

real    0m12.266s
user    0m0.030s
sys     0m0.075s
```

**Comment by NickGerleman** [comment]
@tido64 what repo were you testing?

**Comment by chrisglein** [comment]
> @chrisglein this seems like a pretty clear regression to build performance. Could you help me understand why you considered this item an enhancement?

Based on the chart it looked like the primary regression was 0.64->0.65. So yes a regression, but not so much of a recent one. What I couldn’t remember is if autolinking also filled in feature gaps in that time period (it was spanned over a couple releases), which could explain the extra time.

The short answer to your question is that if it's a regression it's one we shipped 3 versions ago. There's a point at which fixing the bug is a feature, and at first glance that's what this looked like, but I expected more discussion to happen in triage. Looks like it sorted out over mail.

**Comment by jonthysell** [comment]
For future reference:

The slow performance was not widespread, but due to the file structure of the react-native-test-app, which made the RN CLI's search for files unoptimal. This caused the react-native config to take ~3s. This was fixed upstream here: https://github.com/react-native-community/cli/pull/1537/files#diff-c0da7a0a6634600898d6e5eac8dd1be0564fe37f7999148e6721b08080105405

RNW did have a bug (fixed in the attached PRs) where we were calling that RN code multiple times when we meant to only call it once and cache the result. Our cache is now fixed and that fix was backported to 0.66.
